hw/ip/otbn/doc/_index.md

```diff
@@ -604,13 +623,41 @@ When it executes this instruction, OTBN:
   </tbody>
 </table>
 
-## Device Interface Functions (DIFs)
 
-{{< dif_listing "sw/device/lib/dif/dif_otbn.h" >}}
+# Writing OTBN applications {#writing-otbn-applications}
 
-## Register Table
+OTBN applications are (small) pieces of software written in OTBN assembly.
+The full instruction set is described in the [ISA manual]({{< relref "hw/ip/otbn/doc/isa" >}}), and example software is available in the `sw/otbn` directory of the OpenTitan source tree.
 
-{{< registers "hw/ip/otbn/data/otbn.hjson" >}}
+## Toolchain support
+
+OTBN comes with a toolchain consisting of an assembler, a linker, and helper tools such as objdump.
+The toolchain wraps a RV32 GCC toolchain and supports many of its features.
+
+The following tools are available:
+* `otbn-as`: The OTBN assembler.
+* `otbn-ld`: The OTBN linker.
+* `otbn-objdump`: objdump for OTBN.
+
+Other tools from the RV32 toolchain can be used directly, such as objcopy.
+
+## Passing of data between the host CPU and OTBN {#writing-otbn-applications-datapassing}
+
+Passing data between the host CPU and OTBN is done through the data memory (DMEM).
+No standard or required calling convention exists, every application is free to pass data in and out of OTBN in whatever format it finds convenient.
+All data passing must be done as long as OTBN is not running, as indicated by the {{< regref "STATUS.busy" >}} bit; during the OTBN operation both the instruction and the data memory are inaccessible from the host CPU.
+
+## Returning from an application
+
+The software running on OTBN signals completion by executing the [`ECALL`]({{< relref "hw/ip/otbn/doc/isa#ecall" >}}) instruction.
+
+When it executes this instruction, OTBN:
+- Stops fetching and executing instructions.
+- Sets {{< regref "INTR_STATE.done" >}} and clears {{< regref "STATUS.busy" >}}, marking the operation as completed.
+- Writes zero to {{< regref "ERR_CODE" >}}.
```

Comment:
Maybe not for this PR, but I wonder whether we should require SW to clear ERR_CODE. I'm wondering about weird races where something bogus (an FSM glitch, say) happens on the same cycle as we execute ECALL. I think we probably want a nonzero ERR_CODE in that case.

I suppose the other way to look at it is that the ECALL instruction didn't "finish" executing in this case, so the clean description here still works. What do you think?

Reply:
I thought the same, and then deferred that until we have a bit more clarity on the error handling changes, which are ongoing in parallel (There are a couple other things we could/should do for reliability, e.g. check the STATUS.busy bit even after we receive the interrupts, etc.)

Reply:
Yep, sounds sensible to me too.